File: skills/video-production/scripts/batch_report.mjs
#!/usr/bin/env node
/* batch_report.mjs — aggregate verify.mjs JSON reports into one morning summary.
 *
 *   node batch_report.mjs --reports reports/ --out run-log-summary.md
 *
 * The most useful number it produces is the repeat count per check: if the same
 * check failed across several videos, the fault is in the template, not in the
 * videos.
 */
import { readdirSync, readFileSync, writeFileSync } from "node:fs";
import { basename, extname, join } from "node:path";

const argv = process.argv.slice(2);
const opt = (n) => { const i = argv.indexOf("--" + n); return i === -1 ? undefined : argv[i + 1]; };

const dir = opt("reports");
if (!dir) { console.error("--reports is required"); process.exit(2); }

let files = [];
try {
  files = readdirSync(dir).filter((f) => extname(f) === ".json").sort()
    .map((f) => join(dir, f));
} catch { /* reported below */ }
if (!files.length) { console.log(`no reports in ${dir}`); process.exit(0); }

const rows = [], failCounter = new Map(), warnRows = [];
for (const f of files) {
  const stem = basename(f, ".json");
  let r;
  try { r = JSON.parse(readFileSync(f, "utf8")); }
  catch { rows.push([stem, "UNREADABLE", []]); continue; }
  const failed = r.failed || [];
  rows.push([stem, r.passed ? "PASS" : "FAIL", failed]);
  for (const c of failed) failCounter.set(c, (failCounter.get(c) || 0) + 1);
  for (const c of r.checks || []) if (c.warn) warnRows.push([stem, c.name, c.detail || ""]);
}

const passed = rows.filter(([, s]) => s === "PASS").length;
const total = rows.length;
// ties keep insertion order, the way Counter.most_common does
const ranked = [...failCounter.entries()].sort((a, b) => b[1] - a[1]);

const out = [];
out.push("# Batch summary\n");
out.push(`**${passed}/${total} passed**\n`);

out.push("\n## Per video\n");
out.push("| video | status | failed checks |");
out.push("|---|---|---|");
for (const [name, status, failed] of rows) {
  const mark = status === "PASS" ? "✅" : status === "FAIL" ? "❌" : "⚠️";
  out.push(`| ${name} | ${mark} ${status} | ${failed.length ? failed.join(", ") : "—"} |`);
}

if (ranked.length) {
  out.push("\n## Repeat failures — read this first\n");
  out.push("| check | videos affected | verdict |");
  out.push("|---|---|---|");
  for (const [check, n] of ranked)
    out.push(`| ${check} | ${n} | ${n >= 3 ? "**TEMPLATE FAULT — fix the shared component**" : "isolated"} |`);
  const [worst, n] = ranked[0];
  if (n >= 3)
    out.push(`\n> ⛔ \`${worst}\` failed on ${n} videos. That is a shared-component `
      + `fault, not ${n} separate mistakes. Fix it once, then re-render `
      + `only the affected videos.\n`);
}

if (warnRows.length) {
  out.push("\n## Warnings — did not block delivery\n");
  out.push("| video | check | detail |");
  out.push("|---|---|---|");
  for (const [v, c, d] of warnRows.slice(0, 20)) out.push(`| ${v} | ${c} | ${d} |`);
}

out.push("\n## Not covered by any check — needs a human\n");
out.push("- Does the story land?");
out.push("- Does the sweep read as a showcase or as a list?");
out.push("- Is the claim right for this market?");
out.push("- Does the language sound native or translated?");
out.push("- Does each shape say what its caption says?");
out.push("\n*A video that passed every check can still be a weak video. "
  + "The checks prevent failure; they do not create quality.*");

const text = out.join("\n") + "\n";
if (opt("out")) {
  writeFileSync(opt("out"), text, "utf8");
  console.log(`wrote ${opt("out")}  (${passed}/${total} passed)`);
} else {
  console.log(text);
}

File: skills/video-production/scripts/batch_report.py
#!/usr/bin/env python3
"""
batch_report.py — aggregate verify.py JSON reports into one morning summary.

    python batch_report.py --reports reports/ --out run-log-summary.md

The most useful number it produces is the repeat count per check: if the same
check failed across several videos, the fault is in the template, not in the
videos.
"""

import sys
# Windows consoles default to cp1252; the report carries Arabic and arrows.
try:
    sys.stdout.reconfigure(encoding="utf-8")
    sys.stderr.reconfigure(encoding="utf-8")
except Exception:
    pass
import argparse
import json
from collections import Counter
from pathlib import Path


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--reports", required=True, help="directory of *.json reports")
    ap.add_argument("--out", help="write markdown here (default: stdout)")
    a = ap.parse_args()

    files = sorted(Path(a.reports).glob("*.json"))
    if not files:
        print(f"no reports in {a.reports}")
        return

    rows, fail_counter, warn_rows = [], Counter(), []
    for f in files:
        try:
            r = json.loads(f.read_text(encoding="utf-8"))
        except json.JSONDecodeError:
            rows.append((f.stem, "UNREADABLE", []))
            continue
        failed = r.get("failed", [])
        rows.append((f.stem, "PASS" if r.get("passed") else "FAIL", failed))
        fail_counter.update(failed)
        for c in r.get("checks", []):
            if c.get("warn") and not c.get("ok", True):
                warn_rows.append((f.stem, c["name"], c.get("detail", "")))
            elif c.get("warn"):
                warn_rows.append((f.stem, c["name"], c.get("detail", "")))

    passed = sum(1 for _, s, _ in rows if s == "PASS")
    total = len(rows)

    out = []
    out.append("# Batch summary\n")
    out.append(f"**{passed}/{total} passed**\n")

    out.append("\n## Per video\n")
    out.append("| video | status | failed checks |")
    out.append("|---|---|---|")
    for name, status, failed in rows:
        mark = {"PASS": "✅", "FAIL": "❌"}.get(status, "⚠️")
        out.append(f"| {name} | {mark} {status} | {', '.join(failed) if failed else '—'} |")

    if fail_counter:
        out.append("\n## Repeat failures — read this first\n")
        out.append("| check | videos affected | verdict |")
        out.append("|---|---|---|")
        for check, n in fail_counter.most_common():
            verdict = ("**TEMPLATE FAULT — fix the shared component**"
                       if n >= 3 else "isolated")
            out.append(f"| {check} | {n} | {verdict} |")
        worst, n = fail_counter.most_common(1)[0]
        if n >= 3:
            out.append(
                f"\n> ⛔ `{worst}` failed on {n} videos. That is a shared-component "
                f"fault, not {n} separate mistakes. Fix it once, then re-render "
                f"only the affected videos.\n")

    if warn_rows:
        out.append("\n## Warnings — did not block delivery\n")
        out.append("| video | check | detail |")
        out.append("|---|---|---|")
        for v, c, d in warn_rows[:20]:
            out.append(f"| {v} | {c} | {d} |")

    out.append("\n## Not covered by any check — needs a human\n")
    out.append("- Does the story land?")
    out.append("- Does the sweep read as a showcase or as a list?")
    out.append("- Is the claim right for this market?")
    out.append("- Does the language sound native or translated?")
    out.append("- Does each shape say what its caption says?")
    out.append("\n*A video that passed every check can still be a weak video. "
               "The checks prevent failure; they do not create quality.*")

    text = "\n".join(out) + "\n"
    if a.out:
        Path(a.out).write_text(text, encoding="utf-8")
        print(f"wrote {a.out}  ({passed}/{total} passed)")
    else:
        print(text)


if __name__ == "__main__":
    main()
